csharp/src/Ice/OutgoingFrame.cs

```diff
@@ -186,176 +156,104 @@ public CompressionResult CompressPayload()
-
-                    // The first segment of the binary context is immediately after the payload
-                    ArraySegment<byte> segment = Data[PayloadEnd.Segment].Slice(PayloadEnd.Offset);
-                    if (segment.Count > 0)
-                    {
-                        Data.Insert(PayloadEnd.Segment + 1, segment);
-                        if (binaryContextEnd.Segment == PayloadEnd.Segment)
-                        {
-                            binaryContextLastSegmentOffset -= PayloadEnd.Offset;
-                        }
-                    }
-                    // else the binary context already starts with its own segment
-                }
-
-                int start = PayloadStart.Segment;
-
-                if (PayloadStart.Offset > 0)
-                {
-                    // There is non payload bytes in the first payload segment: we move them to their own segment.
-
-                    ArraySegment<byte> segment = Data[PayloadStart.Segment];
-                    Data[PayloadStart.Segment] = segment.Slice(0, PayloadStart.Offset);
-                    start += 1;
-                }
-
-                Data.RemoveRange(start, PayloadEnd.Segment - start + 1);
+                Payload.Clear();
                 offset += (int)memoryStream.Position;
-                Data.Insert(start, new ArraySegment<byte>(compressedData, 0, offset));
-
-                PayloadStart = new OutputStream.Position(start, 0);
-                PayloadEnd = new OutputStream.Position(start, offset);
-                Size = Data.GetByteCount();
-
-                if (_binaryContextOstr != null)
-                {
-                    // Recreate binary context OutputStream
-                    _binaryContextOstr =
-                        new OutputStream(_binaryContextOstr.Encoding,
-                                         Data,
-                                         new OutputStream.Position(Data.Count - 1, binaryContextLastSegmentOffset));
-                }
+                Payload.Add(new ArraySegment<byte>(compressedData, 0, offset));
+                _payloadSize = -1; // reset cached value
 
                 // Rewrite the encapsulation size
                 compressedData.AsSpan(encapsulationOffset, sizeLength).WriteEncapsulationSize(
                     offset - sizeLength - encapsulationOffset,
                     Protocol.GetEncoding());
 
-                _payload = null; // reset cache
+                HasCompressedPayload = true;
 
                 return CompressionResult.Success;
             }
         }
 
+        /// <summary>Gets or builds a combined binary context using InitialBinaryContext and _binaryContextOverride.
+        /// This method is used for colocated calls.</summary>
+        internal IReadOnlyDictionary<int, ReadOnlyMemory<byte>> GetBinaryContext()
+        {
+            if (_binaryContextOverride == null)
+            {
+                return InitialBinaryContext;
```

Comment:
Is it safe to return the InitialBinaryContext here? the incoming and outgoing frames will share the ROM for these entries, you can not release it after sent, probably not very important as this code path is only used for collocated

Reply:
Currently it's safe, but we'll need to be careful when we implement array pooling, and the solution then could be to make a copy of this memory.